### PR TITLE
fix(cohorts): Fix bugs with async deletion of cohorts

### PR DESCRIPTION
**Pull request by neilkakkar**
## Problem

Cohort deletions weren't happening, and there was a potential race condition where deletion might happen before the new version comes in, leading to issues.

Further, our delete query was borked, it could potentially delete everything.

This PR adds tests for the entire async deletion flow, and ensures it works fine.

Ran locally, fired the task, checked it works smoothly.

## Changes

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: posthog/models/cohort/cohort.py [hunk 1]
```diff
@@ -230,6 +231,8 @@ def calculate_people_ch(self, pending_version):
             duration=(time.monotonic() - start_time),
         )
 
+        clear_stale_cohort.delay(self.pk, pending_version)
```
[the remark this answers is not in the document]
Comment (neilkakkar) [reply posthog/models/cohort/cohort.py:234]: makes sense

File: posthog/models/async_deletion/delete_cohorts.py [hunk 2]
```diff
@@ -31,7 +31,9 @@ def process(self, deletions: List[AsyncDeletion]):
     def _verify_by_group(self, deletion_type: int, async_deletions: List[AsyncDeletion]) -> List[AsyncDeletion]:
         if deletion_type == DeletionType.Cohort_stale or deletion_type == DeletionType.Cohort_full:
             cohort_ids_with_data = self._verify_by_column("team_id, cohort_id", async_deletions)
-            return [row for row in async_deletions if (row.team_id, row.key.split("_")[0]) not in cohort_ids_with_data]
+            return [
+                row for row in async_deletions if (row.team_id, int(row.key.split("_")[0])) not in cohort_ids_with_data
```
[the remark this answers is not in the document]
Comment (neilkakkar) [reply posthog/models/async_deletion/delete_cohorts.py:35]: correct, the tuple with different type wouldn't register

**Comment by thmsobrmlr** [comment]
Could this lead to performance issues now that we've accumulated a large number of rows per cohort? I saws 3-4 orders of magnitude more rows than people in the underlying cohort. Might be worth sanity checking by deleting on of the larger cohorts manually.

File: posthog/api/test/__snapshots__/test_cohort.ambr [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+  WHERE (team_id = 2
+         AND cohort_id = '1'
+         AND version < '2')
+  '
+---
+# name: TestCohort.test_async_deletion_of_cohort.12
+  '
+  /* celery:posthog.celery.clickhouse_clear_removed_data */
+  ALTER TABLE cohortpeople
+  DELETE
```
Comment (thmsobrmlr) [inline posthog/api/test/__snapshots__/test_cohort.ambr:88]: `ALTER TABLE ... DELETE...` are old fashioned delete. Do we plan on using lightweight deletes in the future? Anything specific preventing us from doing so now?

Comment (neilkakkar) [reply posthog/api/test/__snapshots__/test_cohort.ambr:88]: Does our CH version now support lightweight deletes? Haven't been keeping up 👀

Comment (neilkakkar) [reply posthog/api/test/__snapshots__/test_cohort.ambr:88]: oh, I see events async deletion is already using lightweight deletes!

Yeah, nothing stopping us from using it now

**Comment by neilkakkar** [comment]
Good shout, moved to lightweight deletes, and will manually delete this cohort 20904, which has a LOT of extra rows

```
SELECT COUNT(person_id) as cnt, cohort_id, version 
from cohortpeople
where team_id = 2 and cohort_id = 20904
group by cohort_id, version
order by version desc
```

image

**Comment by neilkakkar** [comment]
Was pretty smooth on cloud, should be okay^^

**Review by thmsobrmlr — APPROVED** [review]

**Pull request merged**